accord-core/src/main/java/accord/local/Command.java

```diff
@@ -399,158 +525,253 @@ private void onChangeInternal(PartialCommand command)
-                if (notifyListenersOnNoop) notifyListeners();
-                return apply();
+
+            case PreApplied:
+                if (executeRanges(safeStore, executeAt()).intersects(writes().keys, safeStore.commandStore()::hashIntersects))
+                {
+                    logger.trace("{}: applying", txnId());
+                    apply(safeStore);
+                }
+                else
+                {
+                    logger.trace("{}: applying no-op", txnId());
+                    setStatus(Applied);
+                    notifyListeners(safeStore);
+                }
         }
-        return Writes.SUCCESS;
+        return true;
     }
 
     /**
      * @param dependency is either committed or invalidated
+     * @param isInsert true iff this is initial {@code populateWaitingOn} call
+     * @return true iff {@code maybeExecute} might now have a different outcome
      */
-    private void updatePredecessor(PartialCommand dependency)
+    private boolean updatePredecessor(Command dependency, boolean isInsert)
     {
         Preconditions.checkState(dependency.hasBeen(Committed));
         if (dependency.hasBeen(Invalidated))
         {
             logger.trace("{}: {} is invalidated. Stop listening and removing from waiting on commit set.", txnId(), dependency.txnId());
             dependency.removeListener(this);
-            removeWaitingOnCommit(dependency);
+            removeWaitingOnCommit(dependency.txnId()); // TODO (now): this was missing in partial-replication; might be redundant?
+            return true;
         }
         else if (dependency.executeAt().compareTo(executeAt()) > 0)
         {
             // cannot be a predecessor if we execute later
-            logger.trace("{}: {} executes after us. Stop listening.", txnId(), dependency.txnId());
+            logger.trace("{}: {} executes after us. Stop listening and removing from waiting on apply set.", txnId(), dependency.txnId());
+            removeWaitingOn(dependency.txnId(), dependency.executeAt());
             dependency.removeListener(this);
+            return true;
         }
         else if (dependency.hasBeen(Applied))
         {
             logger.trace("{}: {} has been applied. Stop listening and removing from waiting on apply set.", txnId(), dependency.txnId());
-            removeWaitingOnApply(dependency);
+            removeWaitingOn(dependency.txnId(), dependency.executeAt());
             dependency.removeListener(this);
+            return true;
         }
-        else
+        else if (isUnableToExecute())
         {
             logger.trace("{}: adding {} to waiting on apply set.", txnId(), dependency.txnId());
-            addWaitingOnApplyIfAbsent(dependency);
+            addWaitingOnApplyIfAbsent(dependency.txnId(), dependency.executeAt());
+            removeWaitingOnCommit(dependency.txnId());
+            return false;
+        }
+        else if (isInsert)
```

Comment:
isInsert seems a bit finicky. It looks like we rely on the !hasBeen(Committed) check in apply to avoid doing it twice?

And then in the false cases they all have to be hitting earlier in the if ladder?

Reply:
> Why would isUnableToExecute that be a reason to wait for apply and not commit?

At method entry: `Preconditions.checkState(dependency.hasBeen(Committed));`

Basically we should only be invoked on a command that is already committed. So we can't be waiting for this.

> Why can't it happen that the dependency executeAt ends up being later and since we wait on apply it is a circular dependency?

For the same reason, executeAt is fixed. We only ever utilise `executeAt` for this kind of logic once it is committed, as prior to this we have no way of knowing it won't move in another direction.

Reply:
If it's a precondition on method entry the dependency is committed then why are we `addWaitingOnCommit` for the dependency if `isInsert`? We shouldn't be waiting on that particular dependency.

Reply:
This should be resolved now